### PR TITLE
Migrate unit and integration tests from TravisCI to GitHub Actions

### LINKED ISSUE #547: Consolidate CI from Travis -> GitHub Actions
I'm not very familiar with Travis, but I'm pretty used to Actions and have tools (e.g. https://github.com/nektos/act) and knowledge I can't apply to Travis. I think it'd be better for contributors to only have to know one or the other. Thoughts?

### PR DESCRIPTION
**Pull request by mfisher87**
Closes #547 

This PR migrates the current TravisCI behavior into GitHub Actions.

**New feature**: Unit tests run on multiple versions of Python.

TODO:

- [x] @JessicaS11 add credentials currently being used for TravisCI tests to GitHub repository secrets
- [x] @JessicaS11 make new unit and integration test workflows "required" in GH repository settings
- [x] @JessicaS11 clean up required Travis check  in repository settings 
- [x] @JessicaS11 or @mfisher87 update CI badges on readme (currently point to Travis)

> [!NOTE]
>
> This **does not** attempt to implement an approval process for PRs from forks. We don't have that now, so we're not losing anything by dropping Travis in this PR. This is challenging. I'll open another draft PR with that stuff, and we can hopefully figure it out.
>
> See #588

### CODE REVIEW

**Comment by mfisher87** [comment]
I got a bit frustrated trying to get the integration test "secure approval" process working the way I want it to. I'm going to spend some time on the Harmony -> ECS work today to give myself a break. I made a post in Openscapes Slack asking if anyone has ideas what's going wrong in the meantime :) Any tips greatly appreciated!

**Comment by mfisher87** [comment]
I think I'm going to split this PR into two pieces: (1) migrate Travis -> GHA as-is; (2) the approval thing.

(1) alone provides value by reducing the number of unique automation systems we depend on.

File: .github/workflows/integration_test.yml
Comment (mfisher87) [inline .github/workflows/integration_test.yml:31]: @JessicaS11 these are the secrets we need added to this repository :point_up:

File: .github/actions/install-icepyx/action.yml
Comment (JessicaS11) [inline .github/actions/install-icepyx/action.yml:2]:
```suggestion
description: 'Install icepyx and dev dependencies'
```

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [inline .github/workflows/integration_test.yml:3]:
```suggestion
# NOTE: We're just running the tests that require earthdata login here; we
#       don't distinguish between unit and integration tests yet.
```

Is your thought that we'll eventually transition to unit + integration, so it makes sense to rename to those now?

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [inline .github/workflows/integration_test.yml:10]: Given these are slightly "expensive" tests since they download files (and running them for every commit could ultimately influence the icepyx metrics), what are our options for restricting how often they run for a given PR?

File: .github/workflows/unit_test.yml
Comment (JessicaS11) [inline .github/workflows/unit_test.yml:19]: Learning moment (and I know there's an open issue too): why these two versions?

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:3]: Yes, that's my thinking. It's common terminology that is recognizable at a glance to those familiar with it, and also searchable. I'm not attached to this naming, however!

I would like to, if there's time, eventually split our tests up in to "unit" and "integration" directories.

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:10]: These triggers (`branches:`) would be for every merge, not for every commit on PRs. We can customize the `pull_request:` triggers. By default, they run on `open` and `synchronize` (effectively, when the PR is pushed to, not on every commit). The full list is here: https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#pull_request

I think the best way we can limit these runs is by adding `paths:` filters. That would enable e.g. docs changes to not trigger the expensive tests!

Comment (mfisher87) [reply .github/workflows/integration_test.yml:10]: Should we consider this part of this PR, or create an issue to track the work?

File: .github/workflows/unit_test.yml
Comment (mfisher87) [reply .github/workflows/unit_test.yml:19]: I chose the lowest and highest versions of Python that we support. We can test every version we support, but I was trying to balance exhaustiveness with environmental impact. I personally can't think of a problem that would be caught by running our tests for 3.11, but that may just be my ignorance :)

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml]: It auto-resolved this when I accepted your suggestion, so I un-resolved. You can mark resolved if you feel that's the case :)

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:31]: I added EARTHDATA_PASSWORD and CODECOV_TOKEN to the org/repo secrets. EARTHDATA_USERNAME is included as part of a session fixture within `test_behind_NSIDC_API_login.py`.

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:10]: > (effectively, when the PR is pushed to, not on every commit).

Right - forgot this subtlety - but a PR can have a lot of pushes too!

> Should we consider this part of this PR, or create an issue to track the work?

I would definitely consider it part of this PR. I don't want to have a setup running that will trigger so many runs.

File: .github/workflows/unit_test.yml
Comment (JessicaS11) [reply .github/workflows/unit_test.yml:19]: Works for me (and makes complete sense).

```suggestion
        python-version: ["3.9", "3.12"] #NOTE: min and max Python versions supported by icepyx
```

**Comment by JessicaS11** [comment]
> I got a bit frustrated trying to get the integration test "secure approval" process working the way I want it to. The event isn't triggering, and I don't know how (of if it's possible) to debug that.

Is it possible that it's because the action isn't yet in development?

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: I'm fine with the rename, I just wasn't sure if they were "accurate" titles given we haven't separated the tests yet.

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:10]: >I would definitely consider it part of this PR. I don't want to have a setup running that will trigger so many runs.

Don't want to push back too hard, because I am fine doing it however, but I do want to raise that Travis CI behaves this way already:

https://github.com/icesat2py/icepyx/blob/development/.travis.yml#L21-L24

I would lean towards breaking this work into two PRs: (1) migrate Travis -> GH; (2) more efficient. When those PRs are merged into `development`, they will produce two clearer commits and entries in the changelog. I'm OK with immediately following this PR up with step 2 if you're OK with dividing it!

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml]: There are definitely some things I would call "integration tests" that will run in the unit tests job. I didn't want to make this PR huge by actually digging in to that though :)

**Comment by mfisher87** [comment]
>Is it possible that it's because the action isn't yet in development?

I did run in to that initially. I've been testing on my fork, and I did merge into `development` over there, and still not seeing the events I expected!

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: > There are definitely some things I would call "integration tests" that will run in the unit tests job. I didn't want to make this PR huge by actually digging in to that though :)

That's why I was wondering aloud if it's a misnomer to rename before we've done the separation. I don't have strong feelings about it but didn't want to create a more-confusing situation for someone else in the future.

**Comment by mfisher87** [comment]
@JessicaS11 I stealth-edited some TODOs in the description of this PR. Just realized Travis will block merge if it doesn't run :)

image

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:10]: > Travis CI behaves this way already:

Not quite...

>  - name: behind Earthdata
>  if: branch = main OR commit_message =~ nsidc_tests OR type = cron

This isn't running the behind-earthdata tests on the development branch unless "nsidc_tests" is in the commit message. As a result, it only runs for the cron or before releases (when the PR is to main).

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml]: How do you feel about the priority of separating the tests? If we're not going to do it anytime soon, perhaps different names are appropriate. Perhaps just "Tests" and "Behind EDL tests"? :shrug:

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:10]: Good catch, I was being too hasty in my reading! Would you prefer me to make this PR more reflect the Travis setup? What was the cron interval? It's not shown in the config.

File: .github/workflows/integration_test.yml
Comment (mfisher87) [inline .github/workflows/integration_test.yml:30]:
```suggestion
```

Remove username envvar; it's provided as a fixture already.

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:10]: Cron was weekly. I'm okay changing the config (especially if GH Actions has options that allow us to take a better approach). As we're conversing, I'm recalling that the cron was an effort to catch anything that got messed up through work going into development without having these tests run for every push to every PR to that branch. So I'm happy to nix the cron and have some filters that run it more selectively instead (for PRs to development; we'll probably want to keep it as is for main).

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [inline .github/workflows/integration_test.yml:14]:
```suggestion
    name: "Integration Tests"
```

For some reason this is the name GH uses in the actions interfaces... so it would mean we have a bunch of "test" options in the branch rules, which will mean I will have to forever cross reference.

File: .github/workflows/unit_test.yml
Comment (JessicaS11) [inline .github/workflows/unit_test.yml:15]:
```suggestion
  test:
    name: "Unit test"
    runs-on: "ubuntu-latest"
```

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:16]: What do you think of this filter? https://github.com/icesat2py/icepyx/pull/580/commits/098bb985bbb6b551159ddd23409a3adc900a8d37

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: > How do you feel about the priority of separating the tests?

Medium. Honestly, I think I'd prefer to have more tests/better test coverage. I'm guessing that ultimately we might end up with three categories: behind EDL, unit, and integration. So renaming as you suggest "tests" and "behind EDL tests" lets us move forward with later separating out things in "tests" as needed.

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:25]:
```suggestion
    name: "behind EDL test"
```

Comment (mfisher87) [reply .github/workflows/integration_test.yml:25]: Not sure I follow. I'm seeing this, and I was trying to avoid "Integration tests/Integration tests":

image

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml]: I've been thinking of "behind EDL" as a subset of "integration", and using for example [skipif](https://docs.pytest.org/en/stable/how-to/skipping.html#skipif) to control whether those tests run in certain environments. What do you think?

**Comment by mfisher87** [comment]
I was confused why the tests weren't passing, and it turns out they also depend on an envvar named `$NSIDC_LOGIN`. When they are both populated with the password, things go!

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:25]: It's inconsistent. But viewed in the action tab, it only shows "Test". And when I tried to add the tests as branch rules, all I can see is like 5 versions of "test" (with varying capitals and [3.9] or [3.12]). The ones with python versions become the only names I can logically link to a specific test.

image

Comment (mfisher87) [reply .github/workflows/integration_test.yml:25]: Yuck! Thanks, GitHub :laughing:

Comment (mfisher87) [reply .github/workflows/integration_test.yml]: 7bb6cb5

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml:16]: It still feels to me like it will run a lot (and a lot more than necessary)... I'd like to revisit with fresh eyes tomorrow, but what are your thoughts on something like:
```
on:
  push: # these are ultimately the runs that inform our README status badges and provide the "current status"
    branches:
      - "main"
      - "development"
  pull_request: # we want to know that we're not going to break anything with the release
    branches:
      - "main"
 pull_request_review:
    branches:
      - "development"
    paths:
      (list from above)
```

If I've got it right, we run: (1) once for each release (the push to main); (1) one (or more, but hopefully only one) time when we open a release PR from development to main (the pull_request trigger); and (1-2) once or twice for each PR to development (possibly once after it's approved if a relevant path is touched, the pull_request_review trigger, and once on the development branch for current status (the push to development)). By my calculations, this keeps us under ~50 downloads/month, probably with the exception of the next couple!

File: README.rst
Comment (JessicaS11) [inline README.rst:45]: I don't suppose there's a magical way for this to check both action statuses (i.e. we don't want it to show as passing if the non-integration tests fail)?

Comment (JessicaS11) [reply README.rst:45]: There is! https://docs.github.com/en/actions/monitoring-and-troubleshooting-workflows/monitoring-workflows/adding-a-workflow-status-badge#using-the-event-parameter

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml:16]: I've never used `pull_request_review` event before. I think we will only want to trigger on the "submitted" type for that event. Let's try it out! NOTE: This still won't fix our fork problem :)

Comment (JessicaS11) [reply .github/workflows/integration_test.yml:16]: > I think we will only want to trigger on the "submitted" type for that event. Let's try it out! 

Or "approved"? I think this is how I set up the uml diagram action so it only runs once per PR.

> NOTE: This still won't fix our fork problem :)

Wouldn't that be nice...

Comment (mfisher87) [reply .github/workflows/integration_test.yml]: 4183ece look OK?

File: README.rst
Comment (mfisher87) [reply README.rst:45]: There isn't a way that I know of. The document you linked is just for filtering by event type. So we could show a badge for the status of the last integration test caused by a push event to the main branch.

If we want to show a combined badge for integration tests and unit tests, we will need to combine the workflows, to the best of my knowledge.

File: .github/workflows/integration_test.yml
Comment (mfisher87) [reply .github/workflows/integration_test.yml]: >Or "approved"? I think this is how I set up the uml diagram action so it only runs once per PR.

There is no approved event type. That workflow checks the current state of the PR after triggering the workflow, and if it's not "approved", it sets the workflow to "skipped": https://github.com/icesat2py/icepyx/blob/a52bc378ab5817db38cd2b47b5ab0c5352128061/.github/workflows/uml_action.yml#L10

If there was an approved event type, then we could avoid the triggering in the first place :(

But since we have so many other possible triggers we would need something like `${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}`. Thoughts?

File: README.rst
Comment (mfisher87) [reply README.rst:45]: What about a separate badge for each class of tests?

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: Those sound like reasonable approaches (and we can adjust the triggers for the integration tests more broadly then).

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: > There is no approved event type.

Guess that was wishful thinking on my part then.

> Since we have other possible triggers we would need something like ${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}. Otherwise, integration tests would be skipped on our release PRs until they were approved, and if we want that behavior we can remove the pull_request trigger. Thoughts?

I'm happy with the latest edits versus this. I don't think we tend to have a lot of official "reviews" until we're getting pretty close to merging, and it would be great to have the tests run on release PRs right away.

File: README.rst
Comment (JessicaS11) [reply README.rst:45]: I read the docs too literally, leading me to my not-quite-right interpretation!

> If we want to show a combined badge for integration tests and unit tests, we will need to combine the workflows, to the best of my knowledge. With the complexity of our triggers, and the future need to approve integration tests from forks, I don't want to do that 😆

100%. We should not combine the test workflows for badging purposes!

> What about a separate badge for each class of tests?

I think that's probably the way to go.

File: README.rst
Comment (JessicaS11) [inline README.rst:43]:
```suggestion
.. |Unit Tests Status (main)| image:: https://github.com/icesat2py/icepyx/actions/workflows/unit_test.yml/badge.svg?branch=main
    :target: https://github.com/icesat2py/icepyx/actions/workflows/unit_test.yml

.. |Integration Tests Status (main)| image:: https://github.com/icesat2py/icepyx/actions/workflows/integration_test.yml/badge.svg?branch=main
    :target: https://github.com/icesat2py/icepyx/actions/workflows/integration_test.yml
```

File: README.rst
Comment (JessicaS11) [reply README.rst:45]: Lines 10 and 12:
```
Latest release (main branch): |Docs Status main| |Unit Tests Status (main)| |Integration Tests Status (main)| |Code Coverage main|
Current development version (development branch): |Docs Status dev| |Unit Tests Status (dev)| |Integration Tests Status (dev)| |Code Coverage dev| |Pre-commit dev|
```

File: .github/workflows/integration_test.yml
Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: > I'm happy with the latest edits versus this.  I don't think we tend to have a lot of official "reviews" until we're getting pretty close to merging, and it would be great to have the tests run on release PRs right away.

I might need to rescind this. I'm noticing the integration test is running, which makes me think that GH considers a comment (regardless of approval) to be a "review", thus triggering the action for any comment made on the PR.

Comment (JessicaS11) [reply .github/workflows/integration_test.yml]: I'm thinking the least unpalatable option at this point is your suggestion to remove the PR trigger and add something like `${{ github.event.action != "pull_request_review" || github.event.review.state == "approved" }}`.

Comment (mfisher87) [reply .github/workflows/integration_test.yml]: Let's try it on for size!

File: README.rst
Comment (mfisher87) [reply README.rst]: I think we've got this now, please unresolve if you disagree :)

**Review by JessicaS11 — APPROVED** [review]
Thanks for all the problem solving and changes on this one! I'm excited to get CI moved onto GH, so thanks for making it happen, Matt.

**Comment by mfisher87** [comment]
Any time! :rocket:

**Pull request merged**
